subprojects/gdk_rust/gdk_electrum/src/headers/bitcoin.rs

```diff
@@ -0,0 +1,303 @@
+    /// if the file doesn't exist, a chain with only the genesis block (relative to `network`) is returned
+    pub fn new(path: PathBuf, network: Network) -> Result<HeadersChain, Error> {
+        let checkpoints = get_checkpoints(network);
+        if !path.exists() {
+            let last = genesis_block(network).header;
+            let mut file = File::create(&path)?;
+            file.write_all(&serialize(&last))?;
+            let height = 0;
+
+            Ok(HeadersChain {
+                path,
+                height,
+                last,
+                checkpoints,
+            })
+        } else {
+            let mut file = File::open(&path)?;
+            let file_size = file.metadata()?.len();
+
+            file.seek(SeekFrom::Start(file_size - 80))?;
+            let mut buf = [0u8; 80];
+            file.read_exact(&mut buf)?;
+            let height = (file_size as u32 / 80) - 1;
+            let last: BlockHeader = deserialize(&buf)?;
+
+            Ok(HeadersChain {
+                path,
+                height,
+                last,
+                checkpoints,
+            })
+        }
+    }
+
+    pub fn height(&self) -> u32 {
+        self.height
+    }
+
+    pub fn get(&self, height: u32) -> Result<BlockHeader, Error> {
+        let mut file = File::open(&self.path)?;
+        file.seek(SeekFrom::Start(height as u64 * 80))?;
+        let mut buf = [0u8; 80];
+        file.read_exact(&mut buf)?;
+        let header: BlockHeader = deserialize(&buf)?;
+        Ok(header)
+    }
+
+    /// to handle reorgs, it's necessary to remove some of the last headers
+    pub fn remove(&mut self, headers_to_remove: u32) -> Result<(), Error> {
+        if self.height > headers_to_remove {
+            let new_height = self.height - headers_to_remove;
+            let new_size = (new_height + 1) as u64 * 80;
+            let file = OpenOptions::new().write(true).open(&self.path)?;
+            self.last = self.get(new_height)?;
+            self.height = new_height;
+            file.set_len(new_size)?;
+            Ok(())
+        } else {
+            Err(Error::Generic("Cannot remove more headers than what I have".into()))
+        }
+    }
+
+    pub fn tip(&self) -> BlockHeader {
+        self.last
+    }
+
+    /// write new headers to the file if checks are passed
+    pub fn push(&mut self, new_headers: Vec<BlockHeader>) -> Result<(), Error> {
+        let mut serialized = vec![];
+        for new_header in new_headers {
+            let new_height = self.height + 1;
+            if self.last.bitcoin_hash() != new_header.prev_blockhash
+                || new_header.validate_pow(&new_header.target()).is_err()
+            {
+                return Err(Error::InvalidHeaders);
+            }
+
+            if new_height % DIFFCHANGE_INTERVAL == 0 {
+                self.flush(&mut serialized)?;
+                let first = self.get(new_height - DIFFCHANGE_INTERVAL)?;
```

Comment:
Yeah, maybe, how do you imagine the API?

but this is out of the scope of this particular MR since I don't want to depend to a new rust-bitcoin release

Reply:
I was thinking something like `diff_retarget(period_first: &BlockHeader, period_last: &BlockHeader) -> Uint256` if this were to be included in rust-bitcoin.

But it could also be extracted here to a `HeadersChain::diff_retarget(height: u32)` method, to keep `push()` a bit more focused. But It's alright to just keep it as-is too.

Reply:
So, I'll leave it as it is, for now, added a comment https://github.com/Blockstream/gdk/pull/49#issuecomment-681903816 to implement it in the following MT